util/src/main/java/org/datacommons/util/StatChecker.java

```diff
@@ -0,0 +1,351 @@
+    // Only add data points to the counter of the greatest standard deviation that it belongs to.
+    // ie. if the data point is beyond 3 std deviation, only add it to that counter.
+    for (DataPoint dp : timeSeries) {
+      double val = dp.getValues(0).getValue();
+      if (Math.abs(val - stats.mean) > 3 * stats.stdDev) {
+        sigma3Counter.addProblemPoints(dp);
+        continue;
+      }
+      if (Math.abs(val - stats.mean) > 2 * stats.stdDev) {
+        sigma2Counter.addProblemPoints(dp);
+        continue;
+      }
+      if (Math.abs(val - stats.mean) > 1 * stats.stdDev) {
+        sigma1Counter.addProblemPoints(dp);
+      }
+    }
+    for (StatValidationCounter.Builder counter :
+        List.of(sigma1Counter, sigma2Counter, sigma3Counter)) {
+      if (counter.getProblemPointsList().isEmpty()) {
+        continue;
+      }
+      resBuilder.addValidationCounters(counter.build());
+    }
+  }
+
+  private static class Stats {
+    double mean = 0;
+    double stdDev = 0;
+  }
+
+  private static Stats getStats(List<DataPoint> timeSeries) {
+    Stats result = new Stats();
+    if (timeSeries.size() < 2) return result;
+    double weights = 0;
+    double sum = 0;
+    double sumSqDev = 0;
+    for (DataPoint dp : timeSeries) {
+      double val = dp.getValues(0).getValue();
+      if (weights > 0) {
+        sumSqDev += 1 * weights / 1 / (weights + 1) * Math.pow((1 / weights * sum - val), 2);
+      }
+      weights++;
+      sum += val;
+    }
+    if (weights > 0) {
+      result.stdDev = Math.sqrt(sumSqDev * (1.0 / weights));
+      result.mean = sum * (1.0 / weights);
+    }
+    return result;
+  }
+
+  // Goes through sorted (but possibly discontinuous) time series, saving the largest fluctuation,
+  // and the datapoints that cause this fluctuation.
+  // Currently ignore fluctuations starting from 0 (division by 0 problem).
+  protected static void checkPercentFluctuations(
+      List<DataPoint> timeSeries, StatValidationResult.Builder resBuilder) {
+    double maxDelta = 0;
+    DataPoint maxDeltaDP = null;
+    DataPoint maxDeltaBaseDP = null;
+    DataPoint baseDataPoint = null;
+
+    for (DataPoint dp : timeSeries) {
+      // Don't try to compare between times because this is a Sawtooth
+      if (dp.getValuesCount() > 1) return;
+      if (dp.getValuesCount() == 0) continue;
+      double currVal = dp.getValues(0).getValue();
+      if (baseDataPoint != null) {
+        double currDelta;
+        double baseVal = baseDataPoint.getValues(0).getValue();
+        if (baseVal == 0) {
+          currDelta = (currVal) / SMALL_NUMBER;
+        } else {
+          currDelta = (currVal - baseVal) / Math.abs(baseVal);
+        }
+        if (Math.abs(maxDelta) < Math.abs(currDelta)) {
+          maxDelta = currDelta;
+          maxDeltaDP = dp;
+          maxDeltaBaseDP = baseDataPoint;
+        }
+      }
+      baseDataPoint = dp;
+    }
+    PercentDifference.Builder largestPercentDiff =
```

Comment:
We should have a counter for percentdiff > 5x or something?  And effectively put PercentDifference inside StatsValidationCounter proto?

(Feel free to defer this to follow on PR)

Reply:
Yes, I can add counters for percentdiff > 50, 100, 500 and put them inside StatsValidationCounter in a follow up PR. Added a todo.